Balance step in BalancedShardsAllocator for a single shard

abeyad opened the pull request:
This commit introduces a single-shard balance step for deciding on
rebalancing a single shard (without taking any other shards in the
cluster into account).  This method will be used by the cluster
allocation explain API to explain in detail the decision process for
finding a more optimal location for a started shard, if one exists.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +320,99 @@ private void balance() {
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
+                return new RebalanceDecision(canRebalance, Type.NO, "cannot rebalance due to in-flight shard store fetches");
```
maybe we should expand this text to explain why we don't allow balancing while in-flight shard store fetching is going on.

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java:
```diff
@@ -145,6 +145,44 @@ public static ClusterState state(String index, boolean activePrimaryLocal, Shard
     }
 
     /**
+     * Creates cluster state with and index that has #(shardStates) started primary shards and no replicas, using the given number of nodes.
```
I don't understand this, what do you mean by #(shardStates) started primary shards?

abeyad replied:
Fixed the comment to make it more clear

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java:
```diff
@@ -145,6 +145,44 @@ public static ClusterState state(String index, boolean activePrimaryLocal, Shard
+        }
+        discoBuilder.localNodeId(newNode(0).getId());
+        discoBuilder.masterNodeId(newNode(1).getId()); // we need a non-local master to test shard failures
+        final int primaryTerm = 1 + randomInt(200);
```
is setting the primary term required?

abeyad replied:
its not needed, removed

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+    public void testRebalanceNonStartedShardNotAllowed() {
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(),
+            randomFrom(ShardRoutingState.INITIALIZING, ShardRoutingState.UNASSIGNED));
```
what about relocating?

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(),
+            randomFrom(ShardRoutingState.INITIALIZING, ShardRoutingState.UNASSIGNED));
+        ShardRouting unassignedShard = clusterState.routingTable().index("idx").shard(0).primaryShard();
```
the shard is not necessarily unassigned?

abeyad replied:
true, changed variable name to `shard`

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testRebalanceNotAllowed() {
```
testRebalanceNotAllowedByDeciders

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    public void testRebalanceNotAllowed() {
+        AllocationDecider noRebalanceDecider = new AllocationDecider(Settings.EMPTY) {
```
`noAllocationDeciders` and `throttleAllocationDeciders` can be used instead here

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        AllocationDecider canAllocateDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                return Decision.YES;
```
`yesAllocationDeciders`

abeyad replied:
The issue is that `yesAllocationDeciders` returns an instance of `AllocationDeciders`, and I needed to create my own to add a decider that overrides `canRebalance`.

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        List<AllocationDecider> allocationDeciders = Arrays.asList(rebalanceDecider, allocationDecider);
+        final int numShards = randomIntBetween(8, 13);
+        final ShardRoutingState[] shardRoutingStates = new ShardRoutingState[numShards];
+        Arrays.fill(shardRoutingStates, ShardRoutingState.STARTED);
```
where is the variable `shardRoutingStates` actually used?

abeyad replied:
not used, i removed it

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testDontBalanceShardThresholdNotMet() {
```
testDontBalanceShardWhenThresholdNotMet

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testBalanceShard() {
```
testRebalancePossible ?

abeyad replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,172 @@
+        assertNotNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testDontBalanceShardCantAllocate() {
```
solely based on names, hard to distinguish from `testRebalanceNotAllowed`

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -127,6 +127,17 @@ public void allocate(RoutingAllocation allocation) {
+     * from the cluster allocation explain API to explain possible rebalancing decisions for a single
+     * shard.
+     */
+    public RebalanceDecision rebalanceShard(final ShardRouting shard, final RoutingAllocation allocation) {
```
good point, done

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +320,99 @@ private void balance() {
+         * optimally balanced cluster.  This method is invoked from the cluster allocation
+         * explain API only.
+         */
+        private RebalanceDecision balanceShard(final ShardRouting shard) {
```
done

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +320,99 @@ private void balance() {
+
+            Decision canRebalance = allocation.deciders().canRebalance(shard, allocation);
+            if (canRebalance.type() != Type.YES) {
+                return new RebalanceDecision(canRebalance, Type.NO, "rebalancing is not allowed");
```
The point here was to basically say the final decision is "rebalancing not allowed", and then each of the individual deciders will convey why that decider said YES/NO in the output.  The alternatives here is to combine all of the NO deciders output into one string to convey as the final explanation, or just pick one of them to use to convey the final explanation.

abeyad commented:
@ywelsch I pushed some commits that address the code review comments and adds tests comparing the results of `BalanceShardsAllocator#allocate` with the single shard balance method.

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+            final float currentWeight = sorter.weight(currentNode);
+            final AllocationDeciders deciders = allocation.deciders();
+            final String idxName = shard.getIndexName();
+            Map<String, NodeRebalanceDecision> nodeDecisions = new HashMap<>();
```
minor optimization, but we can do `new HashMap<>(modelNodes.length - 1);` to avoid hashmap resizing (if you want)

abeyad replied:
good catch!

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+                    final float currentDelta = absDelta(nodeWeight, currentWeight);
+                    deltaAboveThreshold = lessThan(currentDelta, threshold) == false;
+                    weightWithShardAdded = weight.weightShardAdded(this, node, idxName);
+                    final float proposedDelta = weightWithShardAdded - weight.weightShardRemoved(this, currentNode, idxName);
```
I feel like these methods (the comparisons, deltas, etc), would benefit from some comments explaining what's happening for people reading this code

abeyad replied:
Good idea, will add that

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+                } else {
+                    rebalanceConditionsMet = false;
+                    deltaAboveThreshold = false;
+                    weightWithShardAdded = Float.POSITIVE_INFINITY;
```
I think it's a little strange to define these three variables as `final` in this for loop, then potentially set them only once, then use them in the map `.put` call, instead I think it'd be better to define them before the `if` check as:

```java
boolean rebalanceConditionsMet = false;
boolean deltaAboveThreshold = false;
float weightWithShardAdded = Float.POSITIVE_INFINITY;
```

and then you can remove this `else` branch entirely.

abeyad replied:
done

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+
+            return RebalanceDecision.decision(
+                canRebalance, rebalanceDecisionType, assignedNodeId, nodeDecisions, currentWeight, threshold
+            );
```
This syntax is a little funky, it feels like bracket syntax, but for a function call.

abeyad replied:
changed it

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -1250,6 +1350,152 @@ public boolean move() {
+     */
+    public static final class RebalanceDecision extends RelocationDecision {
+        /** a constant representing no decision taken */
+        public static final RebalanceDecision NOT_TAKEN = new RebalanceDecision(null, null, null, null, null, Float.POSITIVE_INFINITY);
```
Does this need to call the `.decision(...)` static method so an explanation can be produces with the `produceFinalExplanation` method?

abeyad replied:
Do you mean adding an explanation for a NOT_TAKEN decision?  That's possible, though these objects are essentially used to check if the allocation method in question (e.g. `BalancedShardsAllocator#decideRebalance`) even took a decision on the shard at all or it was outside of its realm of responsibility so a different allocator should make an attempt on it.

dakrone replied:
I think it would be nice if it added an explanation that explained it was out of the realm of responsibility, but I'm +0 on adding it, so whatever you'd prefer.

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -1250,6 +1350,152 @@ public boolean move() {
+        /**
+         * Returns the decision for being allowed to rebalance the shard.
+         */
+        public Decision getCanRebalanceDecision() {
```
Should these be annotated with `@Nullable`?

abeyad replied:
done

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/Decision.java:
```diff
@@ -316,6 +321,13 @@ public String label() {
+        @Nullable
+        public String getExplanation() {
+            // multi-decisions have no explanation
+            return null;
```
Should multi-decisions not be a concatenation of their sub-decision explanations?

dakrone commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,235 @@
+        routingAllocation = newRoutingAllocation(new AllocationDeciders(Settings.EMPTY, allocationDeciders), clusterState);
+        ShardRouting shard = clusterState.getRoutingNodes().activePrimary(shardToRebalance.shardId());
+        RebalanceDecision rebalanceDecision = allocator.decideRebalance(shard, routingAllocation);
+        assertEquals(shardToRebalance.relocatingNodeId(), rebalanceDecision.getAssignedNodeId());
```
Since this adds a single node, won't the decision for where to rebalance the shard always be the same? Should the test have multiple nodes so it can ensure it's not just because it's the only node that it can go to?

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -127,6 +127,18 @@ public void allocate(RoutingAllocation allocation) {
+     * shard.
+     */
+    public RebalanceDecision decideRebalance(final ShardRouting shard, final RoutingAllocation allocation) {
+        allocation.debugDecision(true);
```
Good point, I changed it

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+                    if ((subDecision.type() == Type.NO && (explanation == null || explanationCause == Type.THROTTLE))
+                            || (subDecision.type() == Type.THROTTLE && explanation == null)) {
+                        explanation = subDecision.getExplanation();
+                        explanationCause = subDecision.type();
```
Yes, but if our first decision is THROTTLE, we wait to see if there is a NO decision as that is stronger.  However, once we have an explanation based on a NO decision, we can break, so I will add that check.

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +321,104 @@ private void balance() {
+                return new RebalanceDecision(canRebalance, Type.NO, explanation);
+            }
+
+            if (allocation.hasPendingAsyncFetch()) {
```
I wanted the `canRebalance` decision go into the explanation even if there is a pending async shard fetch, so that's why I put this check after the `canRebalance` check.  But if you think that will confuse a user reading the output, then I can change it so the pending async fetch doesn't contain the `canRebalance` decision

abeyad commented:
@ywelsch @dakrone I pushed a few commits that addresses the code review comments

abeyad commented:
retest this please

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java:
```diff
@@ -151,13 +151,13 @@ public static ClusterState state(String index, boolean activePrimaryLocal, Shard
         }
         discoBuilder.localNodeId(newNode(0).getId());
-        discoBuilder.masterNodeId(newNode(1).getId()); // we need a non-local master to test shard failures
+        discoBuilder.masterNodeId(newNode(0).getId());
```
why is this change ok?

abeyad commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java:
```diff
@@ -103,28 +103,33 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
                 return allocation.decision(Decision.YES, NAME, "all allocations are allowed");
             case NONE:
-                return allocation.decision(Decision.NO, NAME, "no allocations are allowed");
+                return allocation.decision(Decision.NO, NAME, "no allocations are allowed due to " + setting(enable, usedIndexSetting));
```
fixed

abeyad commented:
@ywelsch thanks for the review! 

Regarding the comment on why we use 2 shards only in the tests, I added an improved comment in afbd598dcf4b60579afdaf8edd8344faf9f7f121.  

Regarding your question as to why we can assign the master node to the local node here: https://github.com/elastic/elasticsearch/pull/21103/files#diff-7a6367b64dc7e2472730337f4751888dR160, the reason is because we can have only 1 node so in that case local must be master.  However, to make it more general, I've just allowed master to be assigned randomly from any one of the nodes: 6c6f386370c7c3b44f18864850b9341a89ecd979

I also added simulating filter deciding on canAllocate in the tests, to ensure that single-shard balance adheres to the same `canAllocate` decisions as production balance, see f40e0e48fb9602813a4fd9370d89d8325c8b9dc0

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -309,6 +328,109 @@ private void balance() {
+
+
+            if (canRebalance.type() != Type.YES) {
+                return new RebalanceDecision(canRebalance, Type.NO, "rebalancing is not allowed", null, nodeDecisions, currentWeight);
```
If `canRebalance.type()` returns `Type.THROTTLE`, should this pass `Type.THROTTLE` instead of hardcoding `Type.NO`?

abeyad replied:
I checked - `canRebalance` can't return a THROTTLE decision right now...but who is to say that can't change in the future, so your suggestion is safer, I'll change it accordingly.

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -1250,6 +1362,147 @@ public boolean move() {
+                                    String assignedNodeId, Map<String, NodeRebalanceDecision> nodeDecisions, float currentWeight) {
+            super(finalDecision, finalExplanation, assignedNodeId);
+            this.canRebalanceDecision = canRebalanceDecision;
+            this.nodeDecisions = nodeDecisions != null ? Collections.unmodifiableMap(nodeDecisions) : null;
```
Do you think we should keep `null` as the identifier here, or can we use an empty map instead? It'd be nice if we could so we'd avoid NPEs in the future; what do you think?

dakrone commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:
```diff
@@ -1250,6 +1362,147 @@ public boolean move() {
+                    finalExplanation = "throttle moving shard to node [" + assignedNodeId + "], as it is " +
+                                           "currently busy with other shard relocations";
+                } else {
+                    finalExplanation = "moving shard to node [" + assignedNodeId + "] to form a more balanced cluster";
```
It could totally be subsequent work, but I think it'd be great if we could expose the deltas in the explanation also (I mean, for us diagnosing allocation issues it'd be nice to have)

dakrone replied:
Maybe we already expose this and I don't see it?

abeyad replied:
Those are captured in the `NodeRebalanceDecision` for each node - each node will report the deltas, the simulated added weights, etc. so if we chose to move to node `A`, we could look up the node decision for `A` to see all those details.

dakrone commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,273 @@
+        ShardRouting shard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RebalanceDecision rebalanceDecision = allocator.decideRebalance(shard, newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.emptyList()), clusterState));
+        assertSame(RebalanceDecision.NOT_TAKEN, rebalanceDecision);
```
Should this assert that it was not taken due to not being started, rather than not taken due to a different reason?

dakrone replied:
I guess that's not encapsulated into the message, so it's hard to check huh..

abeyad replied:
The issue with adding a message is that I wanted there to be one NOT_TAKEN object so we don't keep allocating different NOT_TAKEN objects for each scenario where a decision is not taken, and in this case, the message isn't so important because the idea is that if one allocator returns a NOT_TAKEN decision, it means its the job of another allocator to decide on that shard.

dakrone commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java:
```diff
@@ -0,0 +1,273 @@
+        RebalanceDecision rebalanceDecision = rebalance.v2();
+        assertEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertNotNull(rebalanceDecision.getFinalExplanation());
```
Maybe there's a better explanation we can assert rather than just checking it has one?

abeyad replied:
Good point, I added a better check against the actual string here

abeyad commented:
@dakrone thank you for the review!

I pushed 0daa6646c9412ff135f5e3537884da325597a0b9 and 89096266b6098860c3c6a48b3ac3a585e1e134eb to address your code review comments, and I replied to the other comments.

abeyad commented:
5.x commit: edfcba24d671952512132f9162ddadfba5ff4ade
